transport.go

```diff
@@ -1,166 +0,0 @@
-	dtnetwork "github.com/filecoin-project/go-data-transfer/network"
-	gstransport "github.com/filecoin-project/go-data-transfer/transport/graphsync"
-	"github.com/ipfs/go-cid"
-	"github.com/ipfs/go-datastore"
-	"github.com/ipfs/go-graphsync"
-	gsimpl "github.com/ipfs/go-graphsync/impl"
-	gsnet "github.com/ipfs/go-graphsync/network"
-	"github.com/ipld/go-ipld-prime"
-	"github.com/libp2p/go-libp2p-core/host"
-	"github.com/libp2p/go-libp2p-core/peer"
-	pubsub "github.com/libp2p/go-libp2p-pubsub"
-	pubsubpb "github.com/libp2p/go-libp2p-pubsub/pb"
-	"github.com/minio/blake2b-simd"
-	errors "golang.org/x/xerrors"
-)
-
-const (
-	// directConnectTicks makes pubsub check it's connected to direct peers every N seconds.
-	directConnectTicks uint64 = 30
-)
-
-// LegTransport wraps all the assets to set-up the data transfer.
-type LegTransport struct {
-	tmpDir string
-	ds     datastore.Batching
-	t      dt.Manager
-	Gs     graphsync.GraphExchange
-	topic  *pubsub.Topic
-
-	refc int32
-}
-
-// MakeLegTransport creates a new datatransfer transport to use with go-legs
-func MakeLegTransport(ctx context.Context, host host.Host, ds datastore.Batching, lsys ipld.LinkSystem, topic string) (*LegTransport, error) {
-	t, err := makePubsub(ctx, host, topic)
-	if err != nil {
-		return nil, err
-	}
-
-	gsnet := gsnet.NewFromLibp2pHost(host)
-	gs := gsimpl.New(context.Background(), gsnet, lsys)
-	tp := gstransport.NewTransport(host.ID(), gs)
-	dtNet := dtnetwork.NewFromLibp2pHost(host)
-
-	// DataTransfer channels use this file to track cidlist of exchanges
-	// NOTE: It needs to be initialized for the datatransfer not to fail, but
-	// it has no other use outside the cidlist, so I don't think it should be
-	// exposed publicly. It's only used for the life of a data transfer.
-	// In the future, once an empty directory is accepted as input, it
-	// this may be removed.
-	tmpDir, err := ioutil.TempDir("", "go-legs")
-	if err != nil {
-		return nil, err
-	}
-	dt, err := datatransfer.NewDataTransfer(ds, tmpDir, dtNet, tp)
-	if err != nil {
-		return nil, err
-	}
-
-	v := &Voucher{}
-	lvr := &VoucherResult{}
-	val := &legsValidator{}
-	if err := dt.RegisterVoucherType(v, val); err != nil {
-		return nil, err
-	}
-	if err := dt.RegisterVoucherResultType(lvr); err != nil {
-		return nil, err
-	}
-	if err := dt.Start(ctx); err != nil {
-		return nil, err
-	}
-
-	return &LegTransport{
-		tmpDir: tmpDir,
-		ds:     ds,
-		t:      dt,
-		Gs:     gs,
-		topic:  t,
-	}, nil
-}
-
-func (lt *LegTransport) addRefc() {
-	atomic.AddInt32(&lt.refc, 1)
-}
-
-// Fetch starts a new transfer with a peer for certain CID using the selector passed as argument.
-//
-// sel determines the selector to use for the exchange
-// Optionally, we can run an onFinish callback when the exchange has finished.
-func (lt *LegTransport) Fetch(ctx context.Context, p peer.ID, c cid.Cid, sel ipld.Node, onFinish func()) (chan cid.Cid, context.CancelFunc, error) {
```

Comment:
I think the `MultiSubscriber` approach is a great idea. The purpose of the `LegTransport` was just to offer an abstraction of pubsub+data-transfer making them transparent for the user. My main concern is that  `LegTransport` included a `Fetch` function to proactively fetch a graph from any peer without requiring an advertisment. This is something we'll eventually need to implement step-by-step syncing in storetheindex. We can probably include this function as part of the multisubscriber interface.

Reply:
@adlrocha why would this function exist here? If you have access to the data transfer (which you do on the current multi-subscriber interface, why not just implement it there. I don't understand what is "go-legs" specific about this -- it appears to be simply a data transfer call

Reply:
i think from the provider perspective the right thing to do will be to limit to the two actually used selector forms: "give me the chain of advertisements back to some cid" and "give me the cids of a specific advertisement.

Reply:
That's a great point, @hannahhoward. The idea was to try an abstract the developer in `go-legs` from vouchers, and the underlying complexity of `go-data-transfer`. I actually like @willscott's idea, the same way that we are currently exposing `LegSelector` we can limit the selectors that can be used within `go-legs` (in the end we'll only potentially use two of them for the subscriber).